chain/chain/src/store.rs

```diff
@@ -923,50 +923,175 @@ impl ChainStoreAccess for ChainStore {
-            StateChangesRequest::SinglePostponedReceiptChanges { account_id, data_id } => {
-                utils::key_for_postponed_receipt_id(account_id, data_id)
+            StateChangesRequest::CodeChanges { account_id } => {
+                utils::KeyForCode::new(account_id).into()
             }
-            StateChangesRequest::AllPostponedReceiptChanges { account_id } => {
-                utils::key_for_all_postponed_receipts(account_id)
+        };
+        storage_key.extend(&data_key);
+
+        let mut changes_iter =
+            self.store.iter_prefix_ser::<RawStateChangesList>(ColKeyValueChanges, &storage_key);
+        let mut exact_key_changes_iter;
+        let changes_iter: &mut dyn Iterator<Item = _> = match state_changes_request {
+            // These request types should have the key matching exactly
+            StateChangesRequest::AccountChanges { .. }
+            | StateChangesRequest::SingleAccessKeyChanges { .. }
+            | StateChangesRequest::CodeChanges { .. } => {
+                exact_key_changes_iter = changes_iter.filter_map(|change| {
+                    let (key, state_changes) = match change {
+                        Ok(change) => change,
+                        error => {
+                            return Some(error);
+                        }
+                    };
+                    if key.len() != storage_key.len() {
+                        None
+                    } else {
+                        debug_assert_eq!(key, storage_key.as_ref() as &[u8]);
+                        Some(Ok((key, state_changes)))
+                    }
+                });
+                &mut exact_key_changes_iter
             }
-        });
-        let common_key_prefix_len = block_hash.as_ref().len()
-            + match state_changes_request {
-                StateChangesRequest::AccountChanges { .. }
-                | StateChangesRequest::SingleAccessKeyChanges { .. }
-                | StateChangesRequest::AllAccessKeyChanges { .. }
-                | StateChangesRequest::CodeChanges { .. }
-                | StateChangesRequest::SinglePostponedReceiptChanges { .. }
-                | StateChangesRequest::AllPostponedReceiptChanges { .. } => storage_key.len(),
-                StateChangesRequest::DataChanges { account_id, .. } => {
-                    utils::key_for_data(account_id, b"").len()
+
+            StateChangesRequest::AllAccessKeyChanges { .. }
+            | StateChangesRequest::DataChanges { .. } => &mut changes_iter,
+        };
+
+        match state_changes_request {
+            StateChangesRequest::AccountChanges { account_id, .. } => {
+                let mut changes = StateChanges::new();
+
+                for change in changes_iter {
+                    let (_, state_changes) = change?;
+                    changes.extend(state_changes.into_iter().map(|(cause, state_change)| {
```

Comment:
I have tried to express it in less code (e.g. I wanted to use `flat_map`, but we want to short-circuit when the iterator fails), and I have also tried to generalize the code to avoid semi-duplication in the matc branches, but it quickly becomes overcomplicated.

Reply:
what is all the added code for?